text/3874-build-std-always.md

```diff
@@ -0,0 +1,1538 @@
+
+## Why not globally cache builds of the standard library?
+[rationale-caching]: #why-not-globally-cache-builds-of-the-standard-library
+
+The standard library is no different than regular dependencies in being able to
+benefit from global caching of dependency builds. It is out-of-scope of this
+proposal to propose a special-cased mechanism for this that applies only to the
+standard library. [cargo#5931] tracks the feature request of intermediate
+artifact caching in Cargo.
+
+↩ [*Caching*][caching]
+
+## Why not link to hosted standard library documentation in generated docs?
+[rationale-generated-docs]: #why-not-link-to-hosted-standard-library-documentation-in-generated-docs
+
+Cargo would need to pass `-Zcrate-attr="doc(html_root_url=..)"` to the standard
+library crates when building them but doesn't have the required information to
+know what url to provide. Cargo would require knowledge of the current toolchain
+channel to build the correct url and doesn't know this.
+
+↩ [*Generated documentation*][generated-documentation]
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+The following are aspects of the proposal which warrant further discussion or
+small details are likely to be bikeshed prior to this part of the RFC's
+acceptance or stabilisation and aren't pertinent to the overall design:
+
+## What should the `build-std` configuration in `.cargo/config` be named?
+[unresolved-config-name]: #what-should-the-build-std-configuration-in-cargoconfig-be-named
+
+What should this configuration option be named? `build-std`?
+`rebuild-standard-library`?
+
+↩ [*Proposal*][proposal]
+
+## What should the "always" and "never" values of `build-std` be named?
+[unresolved-config-values]: #what-should-the-always-and-never-values-of-build-std-be-named
+
+What is the most intuitive name for the values of the `build-std` setting?
+`always`? `manual`? `unconditional`?
+
+`always` combined with the configuration option being named `build-std` -
+`build-std = "always"` - is imperfect as it reads as if the standard library
+will be re-built every time, when it actually just avoids use of the pre-built
+standard library and caches the newly-built standard library.
+
+↩ [*Proposal*][proposal]
+
+## What should `build-std-crates` be named?
+[unresolved-build-std-crate-name]: #what-should-build-std-crates-be-named
+
+What should this configuration option be named?
+
+↩ [*Proposal*][proposal]
+
+## Should the standard library inherit RUSTFLAGS?
+[unresolved-inherit-rustflags]: #should-the-standard-library-inherit-rustflags
+
+Existing designs for *[Opaque dependencies]* intended that `RUSTFLAGS` would not
+apply to the opaque dependency. However, if a target modifier were set using
+`RUSTFLAGS` and build-std ignored the variable, then rustc would fail to build
+the user's project due to incompatible target modifiers. This would necessitate
+that every stable target modifier be exposed via Cargo to be usable in practice.
```

Comment:
I have run into a friction point with the current build-std in that setting `RUSTFLAGS=-Dwarnings` sometimes causes the std build to fail on T3 targets. It may be worth noting somewhere that std's lints will be suppressed and not be affected by RUSTFLAGS, similar to other dependencies.

Reply:
I've mentioned that we'll cap-lints for standard library crates in 4db3d98.